handler/src/main/java/io/netty/handler/ssl/ApplicationProtocolNegotiationHandler.java

```diff
@@ -100,6 +101,14 @@ public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
     public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
         // Let's buffer all data until this handler will be removed from the pipeline.
         bufferedMessages.add(msg);
+        if (!sslHandlerChecked) {
+            sslHandlerChecked = true;
+            if (ctx.pipeline().get(SslHandler.class) == null) {
+                // Just remove ourself if there is no SslHandler in the pipeline and so we would otherwise
+                // buffer forever.
+                removeSelfIfPresent(ctx);
```

Comment:
(_From the other PR https://github.com/netty/netty/pull/11498#issuecomment-884670081_)

> there are other handlers that might put the SslHandler in the pipeline.

I think what you are saying is that some handlers may delay adding an `SslHandler` hence checking the presence of `SslHandler` in `handlerAdded()` may not always succeed for correct pipeline configurations. Did I understand it correctly?

I feel the silent removal of the handler here is confusing and may never surface the issue leading to hard to debug situations. Why not fail and close the channel  instead of silently removing?

Reply:
Shouldn't we also release all the buffered messages to prevent any leaks?

Reply:
These are released in `handlerRemoved()`

Reply:
> (_From the other PR [#11498 (comment)](https://github.com/netty/netty/pull/11498#issuecomment-884670081)_)
> 
> > there are other handlers that might put the SslHandler in the pipeline.
> 
> I think what you are saying is that some handlers may delay adding an `SslHandler` hence checking the presence of `SslHandler` in `handlerAdded()` may not always succeed for correct pipeline configurations. Did I understand it correctly?
> 
> I feel the silent removal of the handler here is confusing and may never surface the issue leading to hard to debug situations. Why not fail and close the channel instead of silently removing?

@NiteshKant this just "mimics" the old behaviour where we would forward the messages. Also for example handlers like `SniHandler` might "delay" the creation of the `SslHandler`